### ISSUE TITLE
[feature request] Add cudaification API for distributions

### ISSUE THREAD

**Issue by Balandat** [pytorch/pytorch#7795]
Right now it's possible to have a distribution "live" on the GPU, so that you can sample to directly get cuda tensors. You'll have to do something like the following though:

    N = Normal(0.0, 1.0)
    N.loc = N.loc.cuda()
    N.scale = N.scale.cuda()

Alternatively, you can directly initialize the parameters on the GPU:

    N_cuda = Normal(torch.tensor([0.0], device='cuda'), torch.tensor([1.0], device='cuda'))

Both of these methods are somewhat verbose and cumbersome. It would be great having something simpler like the following:

        Ncuda = Normal(0.0, 1.0).cuda()

Basically, one would just have to make sure that `.cuda()` gets called on all relevant parameters of the respective distribution. 

Some background on my use case: I'm using distributions to create more abstract notions of priors, which I register inside some Modules. Ideally, I'd just be able to call `.cuda()` on the top Module and have everything moved to GPU, like is the case with most modules - the above feature request is the basis for that.

**Comment by SsnL** [comment]
I'm just pointing out that the recommended way is to use `.to(device)`. Although it would be good for `distributions` to support `to`, and for `modules` to move distributions that are attributes as well.

**Comment by Balandat** [comment]
Note that this would likely require some slight changes to how @lazy_property works (e.g. if we want to be lazy about updating the different representations of the covariance in MultivariateNormal)

**Comment by apaszke** [comment]
cc: @fritzo

**Comment by MaximArtemev** [comment]
Any updates so far?

**Comment by fritzo** [comment]
See #31009. As a workaround, you can construct distribution objects on-the-fly as you need them, and storing their parameters in an `nn.Module`, e.g.
```py
class Thing(nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer("loc", torch.tensor(0.))
        self.register_buffer("scale", torch.tensor(1.))
    def get_dist(self):
        return Normal(self.loc, self.scale)
```

**Comment by albanD** [comment]
We would accept a PR for this.

**Comment by lucaslie** [comment]
I have written a class decorator `@distribution_module` that enables me to "modulify" any `torch.distribution.Distribution` class. 

It simply works by inheriting from the distribution that you want to "modulify" and decorating the new class. Here is an example for `MultivariateNormal`:
```python
@distribution_module
class MultivariateNormalModule(torch.distributions.MultivariateNormal):
    """A "modulified" MultivariateNormal class."""
    pass
```

And here is the decorator: 
```python
def distribution_module(cls):
    """A class decorator to "modulify" torch distributions."""
    # get the actual distribution class from the mro()
    cls_distribution = []
    for parent in cls.mro()[1:]:
        if (
            issubclass(parent, torch.distributions.Distribution)
            and parent != torch.distributions.Distribution
        ):
            cls_distribution.append(parent)
    assert len(cls_distribution) == 1
    cls_distribution = cls_distribution[0]

    class DistributionModule:
        def __init__(self, *args, **kwargs):
            """Initialize distribution and register plain tensors as buffers."""
            super().__init__(*args, **kwargs)

            # after initializing re-register distribution parameters as buffers
            k_tensors = []
            for k in self.__dict__:

                # check if it's a lazy_property
                # in this case we should simply move on instead of evaluating it.
                if self._is_lazy_property(k):
                    continue

                # check if it's a "plain" tensor
                if isinstance(getattr(self, k), torch.Tensor):
                    k_tensors.append(k)

            for k in k_tensors:
                # for a "plain "tensor we will now register it as buffer.
                val = getattr(self, k)
                delattr(self, k)
                self.register_buffer(k, val)

        def __getattribute__(self, name):
            """Return attribute with lazy_property special check."""
            if type(self)._is_lazy_property(name):
                # deleting the attribute from the instance will simply "reset"
                # the lazy_property
                delattr(self, name)
            return super().__getattribute__(name)

        @classmethod
        def _is_lazy_property(cls, name):
            return isinstance(getattr(cls, name, object), lazy_property)

    return type(
        cls.__name__, (DistributionModule, cls_distribution, nn.Module), {}
    )
```

It has worked nicely for me so far although there is certainly a couple of improvements one can do like: 
* In `__getattribute__` instead of simply calling `delattr` one could do a more thorough check to see if the device has changed and only re-evaluate the `lazy_property` if necessary
* Maybe do some recursive checks on sub-modules if distributions inherit from distributions other than the base distribution (`torch.distribution.Distribution`)

**Comment by fritzo** [comment]
This seems like something PyTorch should provide for general Python data structures, not something that should be baked in to every class hierarchy. Why don't we separate concerns, avoid interface bloat, and provide in a single `torch.deep_to()` function that recreates arbitrary Python objects while calling `.to()` on `Tensors` and `Modules` in those objects?

Here's an implementation that works for all Distributions and all Transforms and basically anything else:
```py
class _DeepToMemo(dict):
    def __init__(self, to_args, to_kwargs):
        super().__init__()
        self._to_args = to_args
        self._to_kwargs = to_kwargs

    def get(self, key, default=None):
        result = super().get(key, default)
        if result is default:
            # Assume key is the id of another object, and look up that object.
            old = ctypes.cast(key, ctypes.py_object).value
            if isinstance(old, (torch.Tensor, torch.nn.Module)):  # or maybe duck type here?
                self[key] = result = old.to(*self._to_args, **self._to_kwargs)
        return result

def deep_to(obj, *args, **kwargs):
    memo = _DeepToMemo(args, kwargs)
    return copy.deepcopy(obj, memo)
```

**Comment by francois-rozet** [comment]
> See #31009. As a workaround, you can construct distribution objects on-the-fly as you need them, and storing their parameters in an `nn.Module`, e.g.
> 
> ```python
> class Thing(nn.Module):
>     def __init__(self):
>         super().__init__()
>         self.register_buffer("loc", torch.tensor(0.))
>         self.register_buffer("scale", torch.tensor(1.))
>     def get_dist(self):
>         return Normal(self.loc, self.scale)
> ```

As a generalization of this method, I propose a `Modulify` class that registers (as buffers) all the positional and keyword arguments. The `__call__` method is overloaded to build the object (in our case a distribution). Importantly, the result of `__call__` is cached (`functools.lru_cache`) to build the object only once, but cleared when `_apply` (the sub-method for `.to`, `.cuda` and `.cpu`) gets called.

```python
class Modulify(nn.Module):
    def __init__(
        self,
        constructor: Callable,
        *args,
        **kwargs,
    ):
        super().__init__()

        self.constructor = constructor
        self._args = args
        self._kwargs = kwargs

        for i, arg in enumerate(args):
            if torch.is_tensor(arg):
                self.register_buffer(f'arg_{i}', arg)

        for k, v in kwargs.items():
            if torch.is_tensor(v):
                self.register_buffer(k, v)

    @property
    def args(self) -> list:
        return [
            self.get_buffer(f'arg_{i}') if torch.is_tensor(arg) else arg
            for i, arg in enumerate(self._args)
        ]

    @property
    def kwargs(self) -> dict:
        return {
            k: self.get_buffer(k) if torch.is_tensor(v) else v
            for k, v in self._kwargs.items()
        }

    @lru_cache(None)
    def __call__(self):
        return self.constructor(*self.args, **self.kwargs)

    def _apply(self, *args, **kwargs):  # -> self
        self.__call__.cache_clear()
        return super()._apply(*args, **kwargs)
```

Then, to have a "modulified" normal distribution

```python
dist= Modulify(Normal, loc=torch.tensor(0.), scale=torch,tensor(1.))
x = dist().sample((100,))
```

The strength of this method is that is it very efficient thanks to caching. The main drawback is that it does not work if the constructor creates an internal tensor, unrelated to the arguments.

**Comment by Balandat** [comment]
> The main drawback is that it would not work if the constructor creates an internal tensor unrelated to the arguments. But I don't know if there are such distributions.

This is going to be very inefficient for things like a MVN instantiated with a full covariance matrix. The constructor computes a Cholesky decomposition which is O(n^3) in the dimension of the distribution. With your suggested solution, this would be incurred for both the CPU and the GPU version of the modulified object, whereas the efficient thing to do would be to just move the `scale_tril` matrix to the GPU instead (no compute required).

**Comment by francois-rozet** [comment]
> > The main drawback is that it would not work if the constructor creates an internal tensor unrelated to the arguments. But I don't know if there are such distributions.
> 
> This is going to be very inefficient for things like a MVN instantiated with a full covariance matrix. The constructor computes a Cholesky decomposition which is O(n^3) in the dimension of the distribution. With your suggested solution, this would be incurred for both the CPU and the GPU version of the modulified object, whereas the efficient thing to do would be to just move the `scale_tril` matrix to the GPU instead (no compute required). 

Hum i'm not sure to understand why it is inefficient. Unless you constantly transfer the distribution from `cpu` to `cuda` (which you should not do), the constructor is called only once. Afterwards, it is completely amortized.

**Comment by Balandat** [comment]
The issue is that the constructor does a substantial amount of linear algebra work when given a full covariants matrix. If I understand your implementation correctly calling .to on the wrapper will require the constructor with the new (now moved tensors) in the args / kwargs that is called in obj to still redo that work (now on the gpu). Or am I missing something?

**Comment by francois-rozet** [comment]
You are right about that, but my assumptions are that the user should not call `to` many times. Usually, either the module stays on GPU or on CPU. I guess there might be distributed cases where this does not hold.

The decorator you proposed does not have this issue, but I wonder how it handles the cases were the object creates a list of Tensors internally. Similarly what if the object creates new internal tensors after the initialization ? (for instance for caching)
I guess this could be resolved by overriding the `__setattr__` attribute.

**Comment by francois-rozet** [comment]
I found another solution to enable cudaification for distributions: make them subclasses of `nn.Module` and overwrite the `apply_` method to a `deepapply` function that reclusively finds all tensors referenced in an object. Making `Distribution` a subclass of `nn.Module` is a bit ugly without making a PR (see below), but overwriting `apply_` is very easy.

```python
def deepapply(obj: Any, fn: Callable) -> Any:
    r"""Applies `fn` to all tensors referenced in `obj`"""

    if torch.is_tensor(obj):
        obj = fn(obj)
    elif isinstance(obj, dict):
        for key, value in obj.items():
            obj[key] = deepapply(value, fn)
    elif isinstance(obj, list):
        for i, value in enumerate(obj):
            obj[i] = deepapply(value, fn)
    elif isinstance(obj, tuple):
        obj = tuple(
            deepapply(value, fn)
            for value in obj
        )
    elif hasattr(obj, '__dict__'):
        deepapply(obj.__dict__, fn)

    return obj

__init__ = Distribution.__init__

def init(self, *args, **kwargs):
    __init__(self, *args, **kwargs)

    self.__class__ = type(
        self.__class__.__name__,
        (self.__class__, nn.Module),
        {},
    )

    nn.Module.__init__(self)

Distribution.__init__ = init
Distribution._apply = deepapply
``` 

After that, `.cuda()`, `.cpu()` and `.to(...)` work like a charm on **all** distributions. Most importantly, they are also recognized by other modules as modules and therefore, if we call `.cuda()` on a module containing a distribution, the latter will also be moved.

**Comment by rsilva8** [comment]
@francois-rozet I don't think these solutions work on nested distributions, like when using TransformedDistribution (say, to implement logistic distribution).
